### DISCUSSION TITLE
Error: Undefined values are not allowed - Undefined should map to Null

### DISCUSSION THREAD

**Discussion by e3dio** [porsager/postgres#314] [Ideas]
> Postgres.js won't accept undefined as values in tagged template queries since it becomes ambiguous what to do with the value. If you want to set something to null, use null explicitly.

Undefined values throwing error and preventing successful query is inconvenient and unintuitive to me, if I pass a field with an undefined value I expect it to act as NULL, it seems that would make sense to treat it that way? Can we at least have option to map undefined to NULL so I don't need to check every field? I can't think of a case where it's better to throw error instead of act as NULL. It breaks queries where undefined could be a possible value. `undefined` most clearly maps to the NULL sql value so I think it should do this automatically

```javascript
// undefined fields in object get inserted as NULL

sql`INSERT INTO table ${sql(object)}`
```

https://github.com/porsager/postgres/blob/master/cjs/src/types.js#L79

**Comment by porsager** [comment]
Is it a big inconvenience having to do `someVar || null` or `someVar ?? null` ? 

Or is this with dynamic input - eg `update x set ${ sql({ someVar: undefined }) }` ?

I don't think I'm against adding eg. an `undefinedAsNull: true` option, but some convincing might help? ☺️

**Reply by e3dio** [reply]
It is inconvenient because often there are many fields, in this 15 fields I would need to add `value || null` to every field, where it should really just assume `undefined == null` as a helpful feature rather than an error. When it throws an error the entire query breaks, where probably in all cases a developer would want it to assume a NULL value

**Reply by porsager** [reply]
If I am "a developer" I can tell you that - no, I would assume I made a mistake if I'm not explicitly passing null. That is also the reason for the behaviour, and described in the very text you copied from the documentation. 

In many scenarios it could be a very bad situation if you accidentally null out a value, so the default is not gonna change - but I can see the value in adding the suggested option for those who know what they're doing or don't care about their data.

**Reply by e3dio** [reply]
Also it is more than just inconvenient, there are critical queries on web server that need to be run, if some random input decides to give an undefined value where really `undefined == null` is a true statement and should be treated as null, the entire system breaks because I did not check every field to convert any undefined to null

**Reply by porsager** [reply]
Alright, I'm convinced my choice for this way of handling `undefined` was still the right choice, but I'll be happy to add an option to change the default. Do you have a better idea for the name than `undefinedAsNull`?

**Reply by e3dio** [reply]
That name seems best to me, at least adding the option would help, `undefined == null` is a common mapping and I think should be default over error but if `undefinedAsNull` is option I will use it

**Reply by porsager** [reply]
It's added now and can be used like this in options now `{ transform: { undefined: null } }` . Will be included with the next release

**Reply by aaronyo** [reply]
FWIW, I hit this same issue when dealing with multi row "dynamic" inserts.

When specifying the insert manually, it was indeed not such a big inconvenience to replace with null. But taking advantage of the dynamic sql feels a bit different.

The standard I like for my API and TypeScript definitions is to have optional parameters, rather than  ```<type> | null``` parameters. So this does mean I have to replace undefined when writing to the DB, or use your new "transform" option.

Since I'm not requiring nulls elsewhere in my stack, I don't think it make sense to do so at the DB layer... I prefer to keep things thin and consistent, rather than having transformation between layers.

Anyway, I think what the default should be is a-whole-nother question. I'm happy with the transform config you added.

**Reply by Minigugus** [reply]
@aaronyo Do you suggest using the `DEFAULT` keyword in dynamic insert, when the `transform` option is missing for instance (kind of implicit ```{ transform: { undefined: sql`DEFAULT` } }```)?  I don't see any other solution for **multi row inserts**, as the columns list must be the same for all rows

**Comment by tonyxiao** [comment]
There is no easy way to differentiate between a sql `NULL` vs `jsonb` `null` as is. Both would end up mapping to `null` in the return value. See the table below for example. `data` column is of type `jsonb`

![CleanShot 2024-12-14 at 12 34 47@2x](https://github.com/user-attachments/assets/821aed37-ab67-4485-8466-49a62e1059fb)

What would be the recommended way to handle this in the current paradigm?
